how-to/openldap/ldap-saslauthd-kerberos.md

```diff
@@ -0,0 +1,282 @@
+#
+# Available options in this Debian package:
+# getpwent  -- use the getpwent() library function
+# kerberos5 -- use Kerberos 5
+# pam       -- use PAM
+# rimap     -- use a remote IMAP server
+# shadow    -- use the local shadow password file
+# sasldb    -- use the local sasldb database file
+# ldap      -- use LDAP (configuration is in /etc/saslauthd.conf)
+#
+# Only one option may be used at a time. See the saslauthd man page
+# for more information.
+#
+# Example: MECHANISMS="pam"
+MECHANISMS="kerberos5"
+START=yes
+...
+```
+## Enable and start saslauthd
+Continue by enabling the saslauthd service.
+
+```bash
+sudo systemctl enable saslauthd
+```
+The final step for saslauthd is to start the saslauthd service.
+
+```bash
+sudo systemctl start saslauthd
+```
+## Configure OpenLDAP
+
+The OpenLDAP service needs its own Kerberos principal.
+```bash
+sudo kadmin -p ubuntu/admin
+Authenticating as principal ubuntu/admin with password.
+Password for ubuntu/admin@EXAMPLE.COM
+kadmin: addprinc -randkey ldap/ldap-server.example.com
+No policy specified for ldap/ldap-server.example.com@EXAMPLE.COM; defaulting to no policy
+Principal "ldap/ldap-server.example.com@EXAMPLE.COM" created.
+kadmin: ktadd -k /etc/ldap/ldap.keytab ldap/ldap-server.example.com
+Entry for principal ldap/ldap-server.example.com with kvno 2, encryption type aes256-cts-hmac-sha1-96 added to keytab WRFILE:/etc/ldap/ldap.keytab.
+Entry for principal ldap/ldap-server.example.com with kvno 2, encryption type aes128-cts-hmac-sha1-96 added to keytab WRFILE:/etc/ldap/ldap.keytab.
+kadmin: q
+```
+The permissions of the keytab file need to be secure. Change the group and permissions to allow the slapd process to read the file.
+```bash
+sudo chown root:openldap /etc/ldap/ldap.keytab
+sudo chmod 640 /etc/ldap/ldap.keytab
+```
+OpenLDAP needs to know where to look for the keytab file, what password check method to use, and where to find saslauthd socket. This can be done using the sasl config file for the slapd.
+```bash
+sudo vi usr/lib/sasl2/slapd.conf
+```
+Add this to the file.
+```text
+keytab: /etc/ldap/ldap.keytab
+pwcheck_method: saslauthd
+saslauthd_path: /var/run/saslauthd/mux
+```
+Restart slapd.
+```bash
+sudo systemctl restart slapd
+```
+## Change the user password
+The user we will be testing with is called ubuntu. You can add the user to the OpenLDAP directory by using the ldapadd command. Remember that the ubuntu principal exists in the Kerberos database.
+
+```bash
+ldapadd -x -D cn=admin,dc=example,dc=com -W <<LDIF
+dn: uid=ubuntu,ou=People,dc=example,dc=com
+objectClass: posixAccount
+objectClass: inetOrgPerson
+uid: ubuntu
+cn: ubuntu
+sn: ubuntu
+userPassword: {SASL}ubuntu@EXAMPLE.COM
+uidNumber: 10050
+gidNumber: 10050
+loginShell: /bin/bash
+homeDirectory: /home/ubuntu
+
+dn: cn=ubuntu,ou=Groups,dc=example,dc=com
+objectClass: posixGroup
+cn: ubuntu
+gidNumber: 10050
+memberUid: ubuntu
+LDIF
+```
+
```

Comment:
Can you add a brief explanation/summary here? I think we should point out the path this authentication is taking, and connect all the dots.

First, the simple bind via ldap is looking for the userPassword. But this server is configured via /etc/ldap/sasl2/slapd.conf to use `saslauthd` as the `pwcheck_method`, and the `userPassword` attribute is giving the sasl identity to use. Or maybe the order is different: when `userPassword` has content of the type `{SASL}<id>`, that's when `/etc/ldap/sasl2/slapd.conf` and `pwcheck_method` kicks in, and `saslauthd` is invoked.

`saslauthd` is configured to use `kerberos5` as the verification mechanism, and that invokes all the kerberos bits.

Reply:
I have an idea for an image here, which would also help to show which bits of the process are done in clear text (and vulnerable), and which ones are using encryption (safe). I'll come back to this comment once I draw something.